### PR TITLE
fix(cookies): fallback to isProduction when `baseURL` is not set

### LINKED ISSUE #7156: Cookie "Secure" flag bypassed when baseURL is not configured
### Is this suited for github?

- [x] Yes, this is suited for github

### To Reproduce

  1. Create a Better Auth instance WITHOUT setting `baseURL`:
     ```typescript
     const auth = betterAuth({
       database: drizzleAdapter(db, { provider: 'pg' }),
       secret: process.env.BETTER_AUTH_SECRET,
       // No baseURL configured
     })
     ```

  2. Ensure no `BETTER_AUTH_URL` or `NEXT_PUBLIC_BETTER_AUTH_URL` env vars are set
  3. Deploy to a production environment with `NODE_ENV=production` served over HTTPS
  (e.g., auto-generated CloudFront URL: https://d3e55zkakxzao3.cloudfront.net/)
  4. Sign in and inspect the Set-Cookie response header
  5. Observe: The session cookie is missing the Secure flag

### Current vs. Expected behavior

**Current behavior:**

When `baseURL` is not configured, `getBaseURL()` returns undefined.

https://github.com/better-auth/better-auth/blob/c31c92f3d24f3ccfb6776d4634e0681c6dee2db5/packages/better-auth/src/context/create-context.ts#L105

In `create-context.ts`, this is converted to an empty string:

https://github.com/better-auth/better-auth/blob/c31c92f3d24f3ccfb6776d4634e0681c6dee2db5/packages/better-auth/src/context/create-context.ts#L124

The cookie security logic then checks `"" !== undefined` which is true, so it evaluates `"".startsWith("https://")` which is false:

https://github.com/better-auth/better-auth/blob/8f22ecfd0446eab4c9ba6c5667611a3ba426ed74/packages/better-auth/src/cookies/index.ts#L26-L33

Result: `secure = false`, bypassing the `isProduction` fallback entirely. Cookies are created WITHOUT the Secure flag even when `NODE_ENV=production`.

**Expected behavior:**

When `baseURL` is not configured, the cookie security should fall back to `isProduction`. This enables secure deployments to preview environments with auto-generated HTTPS URLs.

**Discovery**

Setting `baseURL` to `undefined` in `create-context.ts` results in a single test failure (see below). It's unclear whether this test protects nested logic or was written to match the current behavior. Given `baseURL` is typed as `string | undefined`, I would keep the value as `undefined` and patch the nested logic that expects an empty string. 

```
better-auth:test: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
better-auth:test: 
better-auth:test:  FAIL  src/context/create-context.test.ts > base context creation > edge cases > should handle baseURL as undefined
better-auth:test: AssertionError: expected undefined to be '' // Object.is equality
better-auth:test: 
better-auth:test: - Expected: 
better-auth:test: ""
better-auth:test: 
better-auth:test: + Received: 
better-auth:test: undefined
better-auth:test: 
better-auth:test:  ❯ src/context/create-context.test.ts:858:32
better-auth:test:     856|    const getDatabaseType = () => "memory";
better-auth:test:     857|    const res = await createAuthContext(adapter, opts, getDatabaseType);
better-auth:test:     858|    expect(res.options.baseURL).toBe("");
better-auth:test:        |                                ^
better-auth:test:     859|   });
better-auth:test:     860| 
better-auth:test: 
better-auth:test: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
```

### What version of Better Auth are you using?

1.4.10

### System info

```bash
{
  "system": {
    "platform": "darwin",
    "arch": "arm64",
    "version": "Darwin Kernel Version 25.2.0: Tue Nov 18 21:09:45 PST 2025; root:xnu-12377.61.12~1/RELEASE_ARM64_T6030",
    "release": "25.2.0",
    "cpuCount": 11,
    "cpuModel": "Apple M3 Pro",
    "totalMemory": "18.00 GB",
    "freeMemory": "0.28 GB"
  },
  "node": {
    "version": "v22.14.0",
    "env": "development"
  },
  "packageManager": {
    "name": "npm",
    "version": "10.9.2"
  },
  "frameworks": [
    {
      "name": "next",
      "version": "16.0.10"
    },
    {
      "name": "react",
      "version": "19.2.1"
    }
  ],
  "databases": [
    {
      "name": "pg",
      "version": "8.16.3"
    },
    {
      "name": "drizzle",
      "version": "0.44.7"
    },
    {
      "name": "@neondatabase/serverless",
      "version": "1.0.2"
    }
  ],
  "betterAuth": {
    "version": "1.4.10",
    "config": null
  }
}
```

### Which area(s) are affected? (Select all that apply)

Backend

### Auth config (if applicable)

```typescript
import { betterAuth } from "better-auth"
export const auth = betterAuth({
  emailAndPassword: {  
    enabled: true
  },
});
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bytaesu**
docs

---

- Closes https://github.com/better-auth/better-auth/issues/7156

---
## Summary by cubic
Default to secure cookies in production when baseURL is not set. Prevents insecure cookies in prod. Closes #7156.

- **Bug Fixes**
  - Use NODE_ENV=production as fallback to enable Secure cookies when baseURL is missing.
  - Added test to assert Secure is present in Set-Cookie for prod with no baseURL.

<sup>Written for commit bd26c93dd68081f2ffe38e8f10ef2cb30b163620. Summary will update on new commits.</sup>

### CODE REVIEW

File: packages/better-auth/src/cookies/cookies.test.ts [hunk 1]
```diff
@@ -87,6 +87,40 @@ describe("cookies", async () => {
...
+			},
+			{
+				onResponse(context) {
+					const setCookie = context.response.headers.get("set-cookie");
+					expect(setCookie).toContain("Secure");
+				},
+			},
+		);
+
+		// Clean up
```
[the remark this answers is not in the document]
Comment (bytaesu) [reply packages/better-auth/src/cookies/cookies.test.ts]: I don't want to use try–catch in the test code, so I split it into a nested describe block.

File: packages/better-auth/src/cookies/index.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export function createCookieGetter(options: BetterAuthOptions) {
 	const secure =
 		options.advanced?.useSecureCookies !== undefined
 			? options.advanced?.useSecureCookies
-			: options.baseURL !== undefined
+			: options.baseURL
```
Comment (himself65) [inline packages/better-auth/src/cookies/index.ts:31]: What else could it be?

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/better-auth/src/cookies/cookies.test.ts [hunk 3]
```diff
@@ -87,6 +87,43 @@ describe("cookies", async () => {
...
+				baseURL: undefined,
+			});
+
+			await client.signIn.email(
+				{
+					email: testUser.email,
+					password: testUser.password,
+				},
+				{
+					onResponse(context) {
```
Comment (himself65) [inline packages/better-auth/src/cookies/cookies.test.ts:130]: expect this to be called, use `vi.fn`

File: packages/better-auth/src/cookies/index.ts (hunk 2 above)
Comment (bytaesu) [reply packages/better-auth/src/cookies/index.ts:31]: -> ""

https://github.com/better-auth/better-auth/blob/c31c92f3d24f3ccfb6776d4634e0681c6dee2db5/packages/better-auth/src/context/create-context.ts#L124

File: packages/better-auth/src/cookies/cookies.test.ts (hunk 3 above)
Comment (bytaesu) [reply packages/better-auth/src/cookies/cookies.test.ts:130]: If `onResponse` isn’t called, the `expect` line won’t run anyway, so it's fine 🤔

Also, other parts of the current test file use the same approach, so if needed, it's more appropriate to refactor everything at once later rather than in this PR.
